jgrapht-core/src/main/java/org/jgrapht/alg/clustering/ModularityMeasurer.java

```diff
@@ -0,0 +1,174 @@
+     */
+    public ModularityMeasurer(Graph<V, E> graph, boolean validatePartitions)
+    {
+        this.graph = Objects.requireNonNull(graph);
+        this.validatePartitions = validatePartitions;
+        this.inDegrees = new HashMap<>();
+        this.outDegrees = new HashMap<>();
+
+        precomputeDegrees(graph);
+    }
+
+    /**
+     * Compute the modularity of a vertex partition.
+     * 
+     * @param partitions the partitions
+     * @return the modularity
+     */
+    public double modulariry(List<Set<V>> partitions)
+    {
+        if (validatePartitions) {
+            validatePartitions(partitions);
+        }
+
+        boolean isDirected = graph.getType().isDirected();
+        double mod = 0d;
+        for (Set<V> p : partitions) {
+            for (V u : p) {
+                for (V v : p) {
+                    double w = 0d;
+                    for (E e : graph.getAllEdges(u, v)) {
+                        w += graph.getEdgeWeight(e);
+                    }
+                    if (u == v && !isDirected) {
+                        w *= 2d;
+                    }
+                    mod += w - inDegrees.get(u) * outDegrees.get(v) * normalization;
+                }
+            }
+        }
+
+        return mod * normalization;
+    }
+
+    /**
+     * Check that the input is indeed a partition of the vertices.
+     * 
+     * @param partitions the partitions to check
+     */
+    private void validatePartitions(List<Set<V>> partitions)
+    {
+        Set<V> used = new HashSet<>();
+        for (Set<V> p : partitions) {
+            for (V v : p) {
+                if (!used.add(v) || !graph.containsVertex(v)) {
+                    throw new IllegalArgumentException(INVALID_PARTITION_OF_VERTICES);
+                }
+            }
+        }
+        if (used.size() != graph.vertexSet().size()) {
+            throw new IllegalArgumentException(INVALID_PARTITION_OF_VERTICES);
+        }
+    }
+
+    /**
+     * Pre-compute vertex (weighted) degrees and normalization factor.
+     * 
+     * @param graph the input graph
+     */
+    private void precomputeDegrees(Graph<V, E> graph)
+    {
+        GraphType type = graph.getType();
+
+        if (type.isWeighted()) {
+            m = graph.edgeSet().stream().collect(Collectors.summingDouble(graph::getEdgeWeight));
+            for (V v : graph.vertexSet()) {
+                double sum = 0d;
+                for (E e : graph.incomingEdgesOf(v)) {
+                    sum += graph.getEdgeWeight(e);
+                }
+                inDegrees.put(v, sum);
+                sum = 0d;
+                for (E e : graph.outgoingEdgesOf(v)) {
+                    sum += graph.getEdgeWeight(e);
+                }
+                outDegrees.put(v, sum);
+            }
+        } else {
+            m = graph.edgeSet().size();
+            for (V v : graph.vertexSet()) {
+                inDegrees.put(v, Double.valueOf(graph.inDegreeOf(v)));
```

Comment:
`Double.valueOf(...)` seems to be redundant as `int` to `double` conversion is automatic